Implement `getWithRelationship[s]Paged`

feltech opened the pull request:
Closes #46. Implementation developed alongside new API compliance suite tests: https://github.com/OpenAssetIO/OpenAssetIO/pull/1002.

foundrytom commented on tests/fixtures.py:
```diff
@@ -215,4 +215,116 @@
             ),
         },
     },
+    "Test_getWithRelationshipPaged": {
```
[question] How do we feel about the fixture-splosion vs some use of shared fixtures, now we have a few methods. (These fixtures are largely the same right, but very hard to tell from a glance). I wonder if we should factor some of these out to constants? Its quite important that the business logic is shared between all these implementations, so they _should_ use the same fixtures at some level I think (see: https://github.com/OpenAssetIO/OpenAssetIO/pull/1002#discussion_r1252896014)

feltech replied:
Good point, at the very least the repeated values should be factored out to constants, which would add some clarity as to what is shared across fixtures.

feltech replied:
Refactored a bunch of values to be constants in c8b749ec

foundrytom replied:
Awesomenessness - makes it much clearer that the expected results are meant to be coherent 🙏

foundrytom reviewed (commented):
Sorry high-level review due to time pressures, but the implementation looks 👌, not had time to scrutinize the test fixtures.

elliotcmorris commented on plugin/openassetio_manager_bal/BasicAssetLibraryInterface.py:
```diff
@@ -297,3 +358,29 @@ def __handle_exception(exc, idx, error_callback):
+        for page_start in range(0, len(entity_references), page_size):
+            self.__pages.append(entity_references[page_start : page_start + page_size])
+
+    def hasNext(self, _hostSession):
+        return self.__page_num < len(self.__pages) - 1
+
+    def next(self, _hostSession):
+        self.__page_num += 1
+
+    def get(self, _hostSession):
+        return self.__pages[self.__page_num] if self.__page_num < len(self.__pages) else []
```
[consider] You likely considered this but discarded it due to the inconveniences of moving the method, but it would be more representative if these methods had the simulated_delay on them, rather than the pager-creation methods.

foundrytom replied:
Ahh yeah - well remembered - it was meant to be on both wasnt it.

feltech replied:
Yup very good point.

feltech replied:
Added in f3405e96 - somewhat bigger change than I expected

elliotcmorris replied:
:+1:

elliotcmorris reviewed (approved):
Just the one minor thing, didn't look too hard at fixtures as discussed.

feltech commented:
Added f6caf8bd to update fixtures since all relationship methods are tested together due to https://github.com/OpenAssetIO/OpenAssetIO/pull/1002#discussion_r1252896014

Rebased onto #55 (hence aade9fd6) to get tests running.

elliotcmorris commented on tests/bal_functional_behaviour_suite.py:
```diff
@@ -142,6 +144,54 @@ def test_when_getWithRelationshipsPaged_called_then_results_delayed_by_specified
             self.createTestContext(access=Context.Access.kRead),
         )
 
+    def test_when_getWithRelationshipPaged_pager_hasNext_called_then_results_delayed(
+        self,
+    ):
+        for query_latency in self.__simulated_latency_subtests():
+            pager = self.__getWithRelationshipPaged_pager()
+            with self.__assert_simulated_latency_applied(query_latency):
+                pager.hasNext()
+
+    def test_when_getWithRelationshipsPaged_pager_hasNext_called_then_results_delayed(
+        self,
+    ):
+        for query_latency in self.__simulated_latency_subtests():
+            pager = self.__getWithRelationshipsPaged_pager()
+            with self.__assert_simulated_latency_applied(query_latency):
+                pager.hasNext()
```
[Consider] I'd question whether this is the right axis to test, the simulated delay implementation concerns the methods themselves. As this is an internal test of BAL, I think you'd be fine only testing each method via one of the endpoints.

feltech replied:
Good point, was still in apiComplianceSuite mode when writing these tests.  Will refactor to remove the duplication.

feltech replied:
De-duped in 6efdaf97 - how's that?

elliotcmorris commented on tests/bal_functional_behaviour_suite.py:
```diff
@@ -142,6 +144,54 @@ def test_when_getWithRelationshipsPaged_called_then_results_delayed_by_specified
+    def test_when_getWithRelationshipsPaged_pager_hasNext_called_then_results_delayed(
+        self,
+    ):
+        for query_latency in self.__simulated_latency_subtests():
```
[Question] Is it not possible to use `__check_simulated_latencies_without_callbacks` for these methods?

feltech replied:
The problem IIRC was that there's a "setup" step to get the pager (`getWithRelationshipsPaged` call), and we want to ensure it isn't delayed and doesn't pollute the test of the pager.

elliotcmorris replied:
:+1:

elliotcmorris reviewed (approved):
Looks good. Couple of comments, nothing blocking.

feltech commented:
Neeearly forgot release notes: force-pushed an update.
